#32 updated to store user functionality in one package

sLisnychyi opened the pull request:
Connected #14
1. changes:
updated package naming to functional style
created User interface as abstraction
2. fixed:
ignore failed test in progress module
ignore dependency check (tried to ignore some specific, but didn't work)
3. added

vkuchyn commented:
@sLisnychyi  thanks.
1. Please, provide brief comment what was done. Sometimes not the whole scope of the task passes on review to move smaller steps. This will help reviewer to understand what to review
2.. Travis build failed - please, fix it. You can see description from the link below in github check section
3. In description of PR(pull request) always add comment "Connect #<issue number>"

waiting for fixes )

sLisnychyi commented:
1. 
- updated package naming to functional style
- created User interface as abstraction
2. fixed 
- ignore failed test in progress module
- ignore dependency check
3.  ok

vkuchyn commented:
@sLisnychyi Thanks. Few remarks according working with issues
1. Look at pull request description - this is a good practice - follow it in future
2. All the commits messages should start with issue tag like "#14: fixed checks with accordance to qulice". Please - follow this practice in future

vkuchyn commented on gamification/pom.xml:
```diff
@@ -63,12 +51,48 @@
             <artifactId>jetty-server</artifactId>
             <version>${jetty.version}</version>
-            <scope>runtime</scope>
+            <scope>compile</scope>
```
@sLisnychyi is this a problem place with dependency checker? Try one more time scope -> provided

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/CommonUser.java:
```diff
@@ -44,63 +43,41 @@
+     * CommonUser constructor by default.
      */
-    public User() {
+    public CommonUser() {
```
@sLisnychyi does this constructor realy needed?

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/CommonUser.java:
```diff
@@ -138,10 +115,9 @@ public final boolean equals(final Object obj) {
             result = false;
         }
-        final User user = (User) obj;
+        final CommonUser user = (CommonUser) obj;
```
@sLisnychyi use if - else structure
``` 
     final boolean result;
     if (this == obj) {
             result = true;
         } else 
          if (obj == null || getClass() != obj.getClass()) {
              result = false;
          } else ...
```
in this case you avoid double check user `assert user !=null`

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/Service.java:
```diff
@@ -54,11 +53,11 @@
      * @return User
      */
-    User getUser(String uuid);
+    CommonUser getUser(String uuid);
```
@sLisnychyi replace with interface

vkuchyn replied:
@sLisnychyi I think, input parameter is not uuid, probably - id

vkuchyn replied:
@sLisnychyi set, pls return type to interface `User`

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/Service.java:
```diff
@@ -54,11 +53,11 @@
      * @return List of users
      */
-    List<User> getUsers();
+    List<CommonUser> getUsers();
```
@sLisnychyi replace with interface

vkuchyn replied:
@sLisnychyi forgot to replace return type with `List<User>`

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/ServiceImpl.java:
```diff
@@ -46,7 +44,7 @@
+     * UserRepositoryCu.
      */
     @Autowired
     private UserRepository repository;
```
@sLisnychyi replace autowire with constructor param

vkuchyn replied:
@sLisnychyi missed above comment

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/ServiceImpl.java:
```diff
@@ -46,7 +44,7 @@
 
     /**
-     * UserRepository.
+     * UserRepositoryCu.
```
@sLisnychyi what does mean Cu?

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/ServiceImpl.java:
```diff
@@ -57,12 +55,12 @@ public final String createUser(final String username) {
 
     @Override
-    public final User getUser(final String uuid) {
+    public final CommonUser getUser(final String uuid) {
```
@sLisnychyi replace with interface

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/ServiceImpl.java:
```diff
@@ -57,12 +55,12 @@ public final String createUser(final String username) {
 
     @Override
-    public final List<User> getUsers() {
+    public final List<CommonUser> getUsers() {
```
@sLisnychyi replace with interface

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/ServiceImpl.java:
```diff
@@ -46,7 +44,7 @@
 public class ServiceImpl implements Service {
```
@sLisnychyi it is not good practice to use Imp for implementation. At least if you don't know what for is this class (good question. Do we really need this Service? What logic is encapsulated here ? ) - use Default.

vkuchyn replied:
@sLisnychyi Service is tooooo abstract name as for class as for interface - need to change to something more concrete (if will decide to leave it)

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/UserControllerImpl.java:
```diff
@@ -37,8 +37,6 @@
-import juja.microservices.gamification.service.Service;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
```
@sLisnychyi There is no interface for COntroller, so suffix Impl not necessary

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/UserRepositoryCustom.java:
```diff
@@ -29,10 +29,9 @@
  */
 
-package juja.microservices.gamification.model.repository;
+package juja.microservices.gamification.user;
```
@sLisnychyi it is not clear what is difference between UserRepository and UserRepositoryCustom? One of them should be deleted

sLisnychyi replied:
@vkuchyn such custom implementation and naming is using by spring-data. 
Firstly we create custom implementation (without word 'custom' there would be exceptions)
Secondly let standard repository interface extend the custom one. Doing so combines the CRUD and custom functionality and makes it available to clients. 

http://stackoverflow.com/questions/17008947/whats-the-difference-between-spring-datas-mongotemplate-and-mongorepository

btw. later i am planning to refactor to use another way of opertation with mongo. for example using  Morphia lib

vkuchyn replied:
@sLisnychyi Have a look [here](https://github.com/vkuchyn/github-issue-stat/blob/master/src/main/java/juja/github/stat/mongo/TimeReportMongoDao.java)
I am not a big fan of repositories because they make us use interface as part of implementation and hide lot of stuff behind though make design not flexible. Even if so, we need to separate spring data interface from our business interface and Custom is not a good name. Better to name domain interface UserRepository. And SD - MongoUserRepository. Like so. But better to use only MongoTemplate and implement logic manually.

vkuchyn replied:
@sLisnychyi remove pls `UserRepositoryCustom`. I suppose - it is not required any more

vkuchyn commented on gamification/src/main/java/juja/microservices/gamification/user/UserRepository.java:
```diff
@@ -42,7 +41,7 @@
  */
 public interface UserRepository
-    extends MongoRepository<User, String>, UserRepositoryCustom {
+    extends MongoRepository<CommonUser, String>, UserRepositoryCustom {
```
@sLisnychyi change implementation to interface

vkuchyn commented on gamification/src/main/webapp/WEB-INF/application-context.xml:
```diff
@@ -12,7 +12,7 @@
     <context:annotation-config/>
     <context:component-scan base-package="juja.microservices.gamification"/>
     <mongo:repositories
-            base-package="juja.microservices.gamification.model.repository"/>
```
@sLisnychyi let's use annotated configuration instead of xml. Please, refactore this point

vkuchyn commented:
@sLisnychyi good job!
I have some remarks - please see comments under code. Fix and notify me with message

vkuchyn commented:
@sLisnychyi please, update changes from upstream master

vkuchyn commented:
@sLisnychyi no any reaction. Please, make fixes according to remarks. Feel free to ask questions

vkuchyn commented:
BTW, I didn't found interfaces Point, Comment what was task goal

sLisnychyi commented:
@vkuchyn done

vkuchyn commented:
@sLisnychyi nice, I'll try to review ASAP (on weekends I am leaving for vacation so max date - next Tuesday)

vkuchyn commented:
@sLisnychyi Well, it is bad practice to work on something without ticket. Even if we talked via skype, phone etc - create ticket by yourself or ask me to do that. Also, sometimes possible to do refactore in scope of other ticket, so as for me this can be connected to #14, because without refactore you cannot go on and block with it.

sLisnychyi commented:
@vkuchyn ok, thanks for explanation

vkuchyn commented:
@sLisnychyi Thanks for fixes. Still few moments left to improve:
1. https://github.com/JuniorsJava/microservices/pull/32#discussion_r70870274
2. https://github.com/JuniorsJava/microservices/pull/32#discussion_r70870696
3. https://github.com/JuniorsJava/microservices/pull/32#discussion_r69279786
4. https://github.com/JuniorsJava/microservices/pull/32#discussion_r69279870
5. https://github.com/JuniorsJava/microservices/pull/32#discussion_r69279881
6. https://github.com/JuniorsJava/microservices/pull/32#discussion_r70871652

vkuchyn commented:
@sLisnychyi how is it going?

vkuchyn commented:
@rultor merge

rultor commented:
> @rultor merge

@vkuchyn OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/9151-235188564)

rultor commented:
> @rultor merge

@vkuchyn Done! FYI, the full log is [here](http://www.rultor.com/t/9151-235188564) (took me 4min)
